sklearn/calibration.py

```diff
@@ -377,145 +417,231 @@ def _more_tags(self):
+
+    test : ndarray, shape (n_test_indicies,)
+        Indices of the testing subset.
+
+    supports_sw : bool
+        Whether or not the `estimator` supports sample weights.
+
+    method : {'sigmoid', 'isotonic'}
+        Method to use for calibration.
+
+    classes : ndarray, shape (n_classes,)
+        The target classes.
+
+    sample_weight : array-like, default=None
+        Sample weights for `X`.
+
+    Returns
+    -------
+    calibrated_classifier : _CalibratedClassiferPipeline instance
+    """
+    if sample_weight is not None and supports_sw:
+        estimator.fit(X[train], y[train],
+                      sample_weight=sample_weight[train])
+    else:
+        estimator.fit(X[train], y[train])
+
+    n_classes = len(classes)
+    pred_method = _get_prediction_method(estimator)
+    preds = _get_predictions(
+        pred_method, pred_method.__name__, X[test], n_classes
+    )
+
+    sw = None if sample_weight is None else sample_weight[test]
+    calibrated_classifier = _fit_calibrator(
+        estimator, preds, y[test], classes, method, sample_weight=sw
+    )
+    return calibrated_classifier
+
 
-    It assumes that base_estimator has already been fit, and trains the
-    calibration on the input set of the fit function. Note that this class
-    should not be used as an estimator directly. Use CalibratedClassifierCV
-    with cv="prefit" instead.
+def _get_prediction_method(clf):
+    """Return prediction method.
+
+    `decision_function` method of `clf` returned, if it
+    exists, otherwise `predict_proba` method returned.
 
     Parameters
     ----------
-    base_estimator : instance BaseEstimator
-        The classifier whose output decision function needs to be calibrated
-        to offer more accurate predict_proba outputs. No default value since
-        it has to be an already fitted estimator.
+    clf : Estimator instance
+        Fitted classifier to obtain the prediction method from.
 
-    method : 'sigmoid' | 'isotonic'
-        The method to use for calibration. Can be 'sigmoid' which
-        corresponds to Platt's method or 'isotonic' which is a
-        non-parametric approach based on isotonic regression.
+    Returns
+    -------
+    prediction_method : callable
+        The prediction method.
+    """
+    if hasattr(clf, "decision_function"):
+        method = getattr(clf, "decision_function")
+    elif hasattr(clf, "predict_proba"):
+        method = getattr(clf, "predict_proba")
+    else:
+        raise RuntimeError("'base_estimator' has no 'decision_function' or "
+                           "'predict_proba' method.")
+    return method
 
-    classes : array-like, shape (n_classes,), optional
-            Contains unique classes used to fit the base estimator.
-            if None, then classes is extracted from the given target values
-            in fit().
 
-    See also
-    --------
-    CalibratedClassifierCV
+def _get_predictions(pred_method, method_name, X, n_classes):
```

Comment:
can't we use `pred_method.__name__` instead of passing `method_name`?

Reply:
You are too quick!
Annoyingly, with:
```
                pred_method = partial(
                    cross_val_predict, this_estimator, X, y, cv=cv,
                    method=method_name
                )
```
you can't get the parameters of the model from `pred_method`. I agree, this is not pretty. Any suggestions?

Reply:
Oh wait I think I can use `signature`.